google/api_core/retry/retry_base.py

```diff
@@ -212,6 +214,12 @@ def _retry_error_helper(
         raise final_exc from source_exc
     if on_error_fn is not None:
         on_error_fn(exc)
+    # next_sleep is fetched after the on_error callback, to allow clients
+    # to update sleep_iterator values dynamically in response to errors
+    try:
+        next_sleep = next(sleep_iterator)
```

Comment:
Please could you add a regression test which requires this new functionality? If this functionality is removed, the tests in this PR still pass.

Reply:
FWIW, there shouldn't be new functionality, this is just moving what used to be [at the top of the loop](https://github.com/googleapis/python-api-core/blob/a5648faf31c30158d59a3127ea37f7c9911ff765/google/api_core/retry/retry_streaming.py#L111) to the bottom

But yeah, I can probably come up with a test that tests the new sequencing of events if you think that would be helpful to have

Reply:
The specific ask is for a test that fails at `main` without the code changes in this PR to avoid regressions after this PR is merged.

Reply:
Yeah sounds good, I'll add a test for the sequencing (it might slip to next week though)